src/bibliography_node.jl

```diff
@@ -0,0 +1,177 @@
+"""Representation of a reference within a [`BibliographyNode`}(@ref).
+
+# Properties
+
+* `anchor_key`: the anchor key for the link target as a string (generally the
+  BibTeX key), or `nothing` if the item is in a non-canonical bibliography
+  block.
+* `label`: `MarkdownAST.Node` for the label of the entry. Usually a simple
+  `Text` node, as labels in the default styles do not have inline formatting.
+    or `nothing` if the style does not use labels
+  the rendered bibliography.
+* `reference`: `MarkdownAST.Node` for the paragraph for the fully rendered
+  reference.
+"""
+struct BibliographyItem
+    anchor_key::Union{Nothing,String}
+    label::Union{Nothing,MarkdownAST.Node{Nothing}}
+    reference::MarkdownAST.Node{Nothing}
+end
+
+
+"""Node in `MarkdownAST` corresponding to a `@bibliography` block.
+
+# Properties
+
+* `list_style`: One of `:dl`, `:ul`, `:ol`, cf. [`bib_html_list_style`](@ref)
+* `canonical`: Whether or not the references in the `@bibliography` block are
+  link targets.
+* `items`: A list of [`BibliographyItem`](@ref) objects, one for each reference
+in the block
+"""
+struct BibliographyNode <: Documenter.AbstractDocumenterBlock
+    list_style::Symbol  # one of :dl, :ul, :ol
+    canonical::Bool
+    items::Vector{BibliographyItem}
+end
+
+
+function Documenter.MDFlatten.mdflatten(io, ::MarkdownAST.Node, b::BibliographyNode)
+    for item in b.items
+        Documenter.MDFlatten.mdflatten(io, item.reference)
+        print(io, "\n\n\n\n")
+    end
+end
+
+
+function Documenter.HTMLWriter.domify(
+    dctx::Documenter.HTMLWriter.DCtx,
+    node::Documenter.Node,
+    bibliography::BibliographyNode
+)
+    @assert node.element === bibliography
+    return domify_bib(dctx, bibliography)
+end
+
+
+function domify_bib(dctx::Documenter.HTMLWriter.DCtx, bibliography::BibliographyNode)
+    Documenter.DOM.@tags dl ul ol li div dt dd
+    list_tag = dl
+    if bibliography.list_style == :ul
+        list_tag = ul
+    elseif bibliography.list_style == :ol
+        list_tag = ol
+    end
+    html_list = list_tag()
+    for item in bibliography.items
+        anchor_key = isnothing(item.anchor_key) ? "" : "#$(item.anchor_key)"
+        html_reference = Documenter.HTMLWriter.domify(dctx, item.reference.children)
+        if bibliography.list_style == :dl
+            html_label = Documenter.HTMLWriter.domify(dctx, item.label.children)
+            push!(html_list.nodes, dt(html_label))
+            push!(html_list.nodes, dd(div["$(anchor_key)"](html_reference)))
```

Comment:
Aren't these `anchor_key` already strings at this point?

Reply:
Yes, thanks! Probably should be `anchor_id`, though, since it has the `#` prepended (making it an HTML id element).